flink-table-store-connector/src/test/java/org/apache/flink/table/store/connector/ReadWriteTableITCase.java

```diff
@@ -19,73 +19,242 @@
 package org.apache.flink.table.store.connector;
 
 import org.apache.flink.api.common.RuntimeExecutionMode;
+import org.apache.flink.configuration.ExecutionOptions;
 import org.apache.flink.table.api.TableResult;
-import org.apache.flink.table.planner.runtime.utils.TestData;
+import org.apache.flink.table.catalog.ObjectIdentifier;
 import org.apache.flink.table.store.file.FileStoreOptions;
 import org.apache.flink.types.Row;
+import org.apache.flink.types.RowKind;
 import org.apache.flink.util.CloseableIterator;
+import org.apache.flink.util.function.TriFunction;
 
+import org.apache.commons.lang3.tuple.Pair;
+import org.assertj.core.api.AbstractThrowableAssert;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.junit.runners.Parameterized;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
-import javax.annotation.Nullable;
-
-import java.math.BigDecimal;
 import java.nio.file.Paths;
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Collections;
+import java.util.LinkedHashMap;
 import java.util.List;
+import java.util.Map;
 import java.util.UUID;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
-import scala.collection.JavaConverters;
-
 import static org.apache.flink.table.planner.factories.TestValuesTableFactory.changelogRow;
 import static org.apache.flink.table.planner.factories.TestValuesTableFactory.registerData;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
-/** IT cases for testing querying managed table dml. */
+/** IT cases for managed table dml. */
 @RunWith(Parameterized.class)
 public class ReadWriteTableITCase extends TableStoreTestBase {
 
-    private final boolean hasPk;
-    @Nullable private final Boolean duplicate;
+    private static final Logger LOG = LoggerFactory.getLogger(ReadWriteTableITCase.class);
+
+    private static final Map<Row, Pair<RowKind, Row>> PROCESSED_RECORDS = new LinkedHashMap<>();
```

Comment:
> Why use a static collection? It is very dangarous for thread safety and memory leak.

Because the concurrent read/write case specs are not added yet. The current cases are restricted to sequential read/write.

Reply:
This is a bad habit, even now without concurrent testing. It makes it difficult to understand the interactions between classes and methods.
Here you can create a new Map, and pass the Map as a parameter to the `update` method.